### PR TITLE
feat(tensorrt yolox): inference and publish mask image from yolox model with semantic segmentation header

### PR DESCRIPTION
**Pull request by badai-nguyen**
## Description

This PR is to add semantic segmentation mask output topic to solve the issue https://github.com/autowarefoundation/autoware.universe/issues/5594

## Related links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-3615)

Related PR https://github.com/autowarefoundation/autoware.universe/pull/5562

## Tests performed
- Test tensorrt_yolox output on internal rosbag:
`ros2 launch tensorrt_yolox yolox_s_plus_opt.launch.xml`
[Screencast from 2024年05月30日 09時02分23秒.webm](https://github.com/autowarefoundation/autoware.universe/assets/94814556/554b5fcb-ddfd-4105-8390-c9ac78ddf194)

- Note that, adding semantic segmentation header causes increasing `processing_time`
  - Tested reults on Embedded Robotic Controller Powered by NVIDIA® Jetson AGX Xavier ™ Module RQX-58G with `preprocess_on_gpu: true`:
  - OS version: 
  - `cat /etc/nv_tegra_release`
  - -> `R32 (release), REVISION: 6.1, GCID: 27863751, BOARD: t186ref, EABI: aarch64, DATE: Mon Jul 26 19:36:31 UTC 2021`

| ros2 launch tensorrt_yolox yolox_s_plus_opt.launch.xml model_name:=yolox-sPlus-T4-960x960-pseudo-finetune | ros2 launch tensorrt_yolox yolox_s_plus_opt.launch.xml model_name:=yolox-sPlus-opt-pseudoV2-T4-960x960-T4-seg16cls |
| --- | --- |
|Average	18.9983671<br /> STD	0.3530808684 | Average	51.28096922 <br />STD	：0.6388894661|
|image | image |  

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by badai-nguyen** [comment]
Since @dan-dnn is still working on update the model so I convert this to draft and reopen when the new version of yolox is ready.

**Comment by badai-nguyen** [comment]
Since the utilized model is reconsidered so I close this for now and reopen it after the model is fixed.

**Comment by badai-nguyen** [comment]
I reopen this since the Yolox-s model with segmantic segmentation is updating at https://github.com/autowarefoundation/autoware/pull/4012

**Review by yukkysaito — APPROVED** [review]
LGTM

**Comment by badai-nguyen** [comment]
@manato @dan-dnn Could you please review this PR as codeowners?

File: perception/tensorrt_yolox/src/tensorrt_yolox.cpp [hunk 1]
```diff
@@ -747,17 +890,26 @@ bool TrtYoloX::feedforward(const std::vector<cv::Mat> & images, ObjectArrays & o
...
+    buffers = {input_d_.get(), out_prob_d_.get(), segmentation_out_prob_d_.get()};
+  }
   trt_common_->enqueueV2(buffers.data(), *stream_, nullptr);
 
   const auto batch_size = images.size();
 
   CHECK_CUDA_ERROR(cudaMemcpyAsync(
     out_prob_h_.get(), out_prob_d_.get(), sizeof(float) * out_elem_num_, cudaMemcpyDeviceToHost,
     *stream_));
+  if (multitask_) {
```
Comment (manato) [inline perception/tensorrt_yolox/src/tensorrt_yolox.cpp:908]:
```suggestion
  if (multitask_ && !use_gpu_preprocess_) {
```

File: perception/tensorrt_yolox/src/tensorrt_yolox.cpp [hunk 2]
```diff
@@ -766,7 +918,42 @@ bool TrtYoloX::feedforwardAndDecode(const std::vector<cv::Mat> & images, ObjectA
     float * batch_prob = out_prob_h_.get() + (i * out_elem_num_per_batch_);
     ObjectArray object_array;
     decodeOutputs(batch_prob, object_array, scales_[i], image_size);
+    // add refine mask using object
     objects.emplace_back(object_array);
+    if (multitask_) {
+      segmentation_masks_.clear();
+      float * segmentation_results =
+        segmentation_out_prob_h_.get() + (i * segmentation_out_elem_num_per_batch_);
```
Comment (manato) [inline perception/tensorrt_yolox/src/tensorrt_yolox.cpp:926]:
```suggestion
    
```

File: perception/tensorrt_yolox/src/tensorrt_yolox.cpp [hunk 3]
```diff
@@ -766,7 +918,42 @@ bool TrtYoloX::feedforwardAndDecode(const std::vector<cv::Mat> & images, ObjectA
...
+          output_dims.d[2] / static_cast<float>(image_size.height));
+        int out_w = static_cast<int>(image_size.width * scale);
+        int out_h = static_cast<int>(image_size.height * scale);
+        cv::Mat mask;
+        if (use_gpu_preprocess_) {
+          float * d_segmentation_results =
+            segmentation_out_prob_d_.get() + (i * segmentation_out_elem_num_per_batch_);
+          mask = getMaskImageGpu(&(d_segmentation_results[counter]), output_dims, out_w, out_h, i);
+        } else {
+          mask = getMaskImage(&(segmentation_results[counter]), output_dims, out_w, out_h);
```
Comment (manato) [inline perception/tensorrt_yolox/src/tensorrt_yolox.cpp:947]:
```suggestion
          float * segmentation_results =
              segmentation_out_prob_h_.get() + (i * segmentation_out_elem_num_per_batch_);
          mask = getMaskImage(&(segmentation_results[counter]), output_dims, out_w, out_h);
```

**Comment by manato** [comment]
@badai-nguyen 
Thank you very much for your contribution. And I'm sorry for the delay in the review of this PR.
Overall, it looks good to me, but I added some suggestions to reduce unnecessary Device to Host memory copy. I'd appreciate it if you could consider them. Thank you!

**Comment by dan-dnn** [comment]
@badai-nguyen
Thank you for your contribution.
I apologize for the delay in reviewing this PR.

Overall, I think it looks good, butI have one suggestion regarding performance.
From the perspective of data transfer speed, it is better to publish the mask data before resizing and let the subscriber resize it.

Also, could you please add the following repository related to trt-YOLOXP in the README?

**Comment by badai-nguyen** [comment]
> @badai-nguyen Thank you very much for your contribution. And I'm sorry for the delay in the review of this PR. Overall, it looks good to me, but I added some suggestions to reduce unnecessary Device to Host memory copy. I'd appreciate it if you could consider them. Thank you!

@manato Thank you for your suggestion, I followed all of them.

**Comment by badai-nguyen** [comment]
> @badai-nguyen Thank you for your contribution. I apologize for the delay in reviewing this PR.
> 
> Overall, I think it looks good, butI have one suggestion regarding performance. From the perspective of data transfer speed, it is better to publish the mask data before resizing and let the subscriber resize it.
> 
> Also, could you please add the following repository related to trt-YOLOXP in the README?

@dan-dnn 
- I updated README here https://github.com/autowarefoundation/autoware.universe/pull/5553/commits/d8006916d058d71e6a256614af5c3208d83aa8f6

- I keep the small side of mask when publishing here  https://github.com/autowarefoundation/autoware.universe/pull/5553/commits/a15ac049be6e0763f28f03fd407bc14af79dc524#diff-c07e477020c17e860704e7e466d42f493fc51a02ea27e8134c203b240079f80fR192-R193

Could you review this again?

**Comment by dan-dnn** [comment]
@badai-nguyen

Thank you for the update.
It looks good for me.

**Review by dan-dnn — APPROVED** [review]

**Review by manato — APPROVED** [review]
@badai-nguyen 
Thank you very much for your consideration. LGMT!

**Pull request merged**